### PR TITLE
Introduce ArrayView::cbegin and ArrayView::cend

### PR DESCRIPTION
**Pull request by masterleinad**
In a recent discussion with @jwitte08 we noticed that `ArrayView::begin()` and `ArrayView::end()` have non-const overloads and that the const overloads return `const_iterator`s.

In the current design, you can't write a range-based for loop that modifies the underlying object for a `const ArrayView` while this is possible in a loop using `operator[]`. Since we tend to pass `ArrayView` objects by `const &`, this inconsistency is a bit unfortunate.

Since a `ArrayView` does not own the memory itself, I would argue that the class should always have `const` behavior, i.e. it should only have `const` attributes and `const` member functions.

To improve the situation and have consistent behavior for `operator[]` and `begin()` this PR makes `begin()` and `end()` const and return (non-const) `iterator`s.
Furthermore, `cbegin()` and `cend()` are introduced that are `const` and return `const_iterator`s.

### CODE REVIEW

**Comment by masterleinad** [comment]
The two failing tests are unrelated:
> 258 - dofs/interpolate_q_system_mixed_01.debug (Failed)
2543 - bits/step-15.debug (Failed)

File: tests/base/array_view_11.cc [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    int i=0;
+    for (auto &&it = arr.cbegin(); it != arr.cend(); ++it, ++i)
+      AssertThrow (*it == i+1, ExcInternalError());
+  }
+
+  {
+    std::vector<int> arr = {0,1,2,3,4,5,6,7,8,9};
+    const ArrayView<int> view (arr);  // writable view
+    for (auto &el: view)
+      ++el;
```
Comment (drwells) [inline tests/base/array_view_11.cc]: I have some reservations about permitting this. I would rather have the `const` overload of `begin()` still return a `const_iterator`: I would prefer to not allow mutating things of type `const ArrayView<T>` in addition to `const ArrayView<const T>`.

**Comment by drwells** [comment]
> Since a `ArrayView` does not own the memory itself, I would argue that the class should always have `const` behavior, i.e it should only have `const` attribute and `const` member functions.

I have to disagree with this: we use `ArrayView` to represent a mutable buffer in a few places in the library, e.g., `Manifold::get_new_points` and `TransfiniteInterpolationManifold::compute_chart_points`.

Otherwise: I agree that we need to support the ranged for loop so I like most of this patch.

**Comment by masterleinad** [comment]
> I have to disagree with this: we use ArrayView to represent a mutable buffer in a few places in the library, e.g., Manifold::get_new_points and TransfiniteInterpolationManifold::compute_chart_points.

But there we don't actually change the `ArrayView` itself (and can't: both `starting_element` and `n_elements` are `const`). I didn't deny that we want to modify the content pointed to.

**Comment by drwells** [comment]
I think that I misunderstood what you originally wrote. I consider modifying the contents at the end of a pointer to be modifying the container itself, i.e., modifying entries of a `const ArrayView<T>` is not ok for the same reasons as modifying a `const std::vector<T>`.

I would rather add a second overload of `operator[]()` that bans modifying `const` array views, but (rereading your posts) you are in favor of this. Does anyone else feel strongly about what behavior we should allow here?

File: tests/base/array_view_11.cc (hunk 1 above)
Comment (masterleinad) [reply tests/base/array_view_11.cc]: The point of this PR is to restore consistency with operator[].
```
  std::vector<int> arr = {0,1,2,3,4,5,6,7,8,9};
  const ArrayView<int> view (arr);  // writable view
  for (unsigned int i=1; i<view.size(); ++i)
    ++(arr[i]);
```
is equivalent to the above code and is already allowed.

**Comment by masterleinad** [comment]
For me it might make sense to allow to move the view and hence declare `starting_element` non-`const`. This is unrelated to the stored type being `const` or not.

In the end, the question is if a `const ArrayView` should imply that the underlying type is `const` or not.

**Comment by masterleinad** [comment]
> I would rather add a second overload of operator[]() that bans modifying const array views, but (rereading your posts) you are in favor of this.

Just as a side-note: This is an incompatible change and breaks our own codebase in many places (probably easily fixable), e.g.
- `source/fe/mapping_manifold.cc:1282`
- `source/fe/fe_poly_tensor.cc:1020`
- `source/fe/mapping_q_generic.cc:3329`

**Comment by drwells** [comment]
Good point. I suspect that we will break something by changing this (i.e., we should leave things the way they are) but I would have preferred to write

```cpp
  virtual
  void
  transform (const ArrayView<const Tensor<1,dim> >                  &input,
             const MappingType                                       type,
             const typename Mapping<dim,spacedim>::InternalDataBase &internal,
             ArrayView<Tensor<1,spacedim> >                          output) const = 0;
```

which is actually the same const-ness and reference-ness as the old `VectorSlice` version. I suspect that we went with this one so that we could get an automatic type conversion from `VectorSlice` to `ArrayView`. @bangerth what do you think?

**Comment by bangerth** [comment]
Can we consider an `ArrayView<T>` the same way as we would a `T*`? To me, an `ArrayView` is basically a pointer that also stores the length of the array pointed to (the latter being immaterial). That means that `ArrayView<T>::begin()` should *always* return a `T*` (or something equivalent) regardless of whether the `ArrayView` object itself is `const` or not. Obviously, then, `ArrayView<const T>::begin()` would return a `const T*`.

I think the patch as is does exactly this. I'm in favor!

**Comment by bangerth** [comment]
And, as a point of completeness, `const ArrayView<T> const_obj(...); const_obj.begin()` would return a `T*` in the same way as one could think of getting a range out of a `T *const p` -- `p` is `const`, but it points to mutable memory.

**Comment by drwells** [comment]
Fair enough; lets keep the current semantics if people are happy with them.

**Comment by drwells** [comment]
Since no one has objected (besides me) lets merge this.

**Pull request merged**